src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LocalProxyVarsGen.java

```diff
@@ -0,0 +1,271 @@
+
+    protected static final Context.Key<LocalProxyVarsGen> valueInitializersKey = new Context.Key<>();
+
+    public static LocalProxyVarsGen instance(Context context) {
+        LocalProxyVarsGen instance = context.get(valueInitializersKey);
+        if (instance == null)
+            instance = new LocalProxyVarsGen(context);
+        return instance;
+    }
+
+    private final Types types;
+    private final Names names;
+    private final Target target;
+    private TreeMaker make;
+    private final UnsetFieldsInfo unsetFieldsInfo;
+    private ClassSymbol currentClass = null;
+    private java.util.List<JCVariableDecl> strictInstanceFields;
+    private Map<JCMethodDecl, Set<Symbol>> strictFieldsReadInPrologue = new HashMap<>();
+
+    private final boolean noLocalProxyVars;
+
+    @SuppressWarnings("this-escape")
+    protected LocalProxyVarsGen(Context context) {
+        context.put(valueInitializersKey, this);
+        make = TreeMaker.instance(context);
+        types = Types.instance(context);
+        names = Names.instance(context);
+        target = Target.instance(context);
+        unsetFieldsInfo = UnsetFieldsInfo.instance(context);
+        Options options = Options.instance(context);
+        noLocalProxyVars = options.isSet("noLocalProxyVars");
+    }
+
+    public void addStrictFieldReadInPrologue(JCMethodDecl constructor, Symbol sym) {
+        Set<Symbol> fieldSet = strictFieldsReadInPrologue.getOrDefault(constructor, new HashSet<>());
+        fieldSet.add(sym);
+        strictFieldsReadInPrologue.put(constructor, fieldSet);
+    }
+
+    public JCTree translateTopLevelClass(JCTree cdef, TreeMaker make) {
+        if (!noLocalProxyVars) {
+            try {
+                this.make = make;
+                return translate(cdef);
+            } finally {
+                // note that recursive invocations of this method fail hard
+                this.make = null;
+            }
+        } else {
+            return cdef;
+        }
+    }
+
+    @Override
+    public void visitClassDef(JCClassDecl tree) {
+        ClassSymbol prevCurrentClass = currentClass;
+        java.util.List<JCVariableDecl> prevStrictInstanceFields = strictInstanceFields;
+        try {
+            currentClass = tree.sym;
+            strictInstanceFields = tree.defs.stream()
+                    .filter(t -> t.hasTag(VARDEF))
+                    .map(t -> (JCVariableDecl)t)
+                    .filter(vd -> vd.sym.isStrict() && !vd.sym.isStatic())
+                    .collect(List.collector());
+            super.visitClassDef(tree);
+        } finally {
+            currentClass = prevCurrentClass;
+            strictInstanceFields = prevStrictInstanceFields;
+        }
+    }
+
+    public void visitMethodDef(JCMethodDecl tree) {
+        if (strictFieldsReadInPrologue.get(tree) != null) {
+            Set<Symbol> fieldSet = strictFieldsReadInPrologue.get(tree);
+            if (!strictInstanceFields.isEmpty()) {
```

Comment:
Do we need this code? After all, we know from `Resolve` that a strict instance field was accessed -- so what is left to verify here?